converter/converter.go

```diff
@@ -0,0 +1,437 @@
+	imports []*proto.Import
+}
+
+// Convert will convert a proto file to a gunk file,
+// saving the file in the same folder as the proto file.
+func Convert(path string, overwrite bool) error {
+	if filepath.Ext(path) != ".proto" {
+		return fmt.Errorf("convert requires a .proto file")
+	}
+	reader, err := os.Open(path)
+	if err != nil {
+		return fmt.Errorf("unable to read file %q: %v", path, err)
+	}
+	defer reader.Close()
+
+	// Parse the proto file.
+	parser := proto.NewParser(reader)
+	d, err := parser.Parse()
+	if err != nil {
+		return fmt.Errorf("unable to parse proto file %q: %v", path, err)
+	}
+
+	// Start converting the proto declarations to gunk.
+	b := builder{}
+	for _, e := range d.Elements {
+		if err := b.handleProtoType(e); err != nil {
+			return fmt.Errorf("error: %v\n", err)
+		}
+		if b.err != nil {
+			return fmt.Errorf("error in %d:%d: %v\n", b.errPos.Line, b.errPos.Column, b.err)
+		}
+	}
+
+	// Convert the proto package and imports to gunk.
+	translatedPkg := b.handlePackage()
+	translatedImports := b.handleImports()
+
+	// Add the converted package and imports, and then
+	// add all the rest of the converted types. This will
+	// keep the order that things were declared.
+	w := &strings.Builder{}
+	w.WriteString(translatedPkg.String())
+	w.WriteString("\n\n")
+	w.WriteString(translatedImports.String())
+	w.WriteString("\n")
+	for _, l := range b.translatedDeclarations {
+		w.WriteString("\n")
+		w.WriteString(l.String())
+		w.WriteString("\n")
+	}
+
+	// TODO: We should run this through the Gunk generator to
+	// make sure that it compiles?
+	// TODO: Run this through the gunk fmt when it is implemented.
+
+	fileToWrite := strings.Replace(filepath.Base(path), ".proto", ".gunk", 1)
+	fullpath := filepath.Dir(path) + "/" + fileToWrite
+
+	if _, err := os.Stat(fullpath); !os.IsNotExist(err) && !overwrite {
+		return fmt.Errorf("path already exists %q, cannot overwrite", fullpath)
+	}
+	if err := ioutil.WriteFile(fullpath, []byte(w.String()), 0644); err != nil {
+		return fmt.Errorf("unable to write to file %q: %v", fullpath, err)
+	}
+
+	return nil
+}
+
+// Write will write output to a string builder, adding in indentation
+// where required. It will write the comment first if there is one,
+// and then write the rest.
+func (b *builder) Write(w *strings.Builder, indent int, comments *proto.Comment, s string, args ...interface{}) {
+	if comments != nil {
+		for _, c := range comments.Lines {
+			for i := 0; i < indent; i++ {
+				fmt.Fprintf(w, "\t")
+			}
+			fmt.Fprintf(w, "//%s\n", c)
+		}
+	}
+	// If we are just writing a comment bail out.
+	if s == "" {
+		return
+	}
+	for i := 0; i < indent; i++ {
+		fmt.Fprintf(w, "\t")
+	}
+	fmt.Fprintf(w, s, args...)
+}
+
+// emitError will save the current error position and error. The error
+// formatting will use these when reporting errors.
+func (b *builder) emitError(pos scanner.Position, s string, args ...interface{}) {
+	b.errPos = pos
+	b.err = fmt.Errorf(s, args...)
```

Comment:
what will happen if we encounter many errors? you should either only save the first error, or ensure that the entire program stops when encountering the first error. You could accomplish the second option with recovered panics, for example.

Reply:
It will bail out with the first error it finds. I didn't think this needed to handle more than one error, at least not to begin with anyway?

Reply:
That's my point - it won't always bail when finding the first error. For example, `handleProtoType` is a function that can emit an error, return a nil error, and the program will continue converting.

The simplest fix would be for emitError to do nothing if b.err is already non-nil. I also find it a bit confusing how some functions can both return an error and emit an error.

Reply:
It should bail when it finds the first error, but I do see your point. Actually, the only purpose of this was to allow for better error messages (with the position). I will change it to return a formatted error rather than this emit error thing, and propagate the formatted error.